annotation/ksp/src/main/kotlin/com/bumptech/glide/annotation/ksp/GlideSymbolProcessor.kt

```diff
@@ -0,0 +1,150 @@
+package com.bumptech.glide.annotation.ksp
+
+import com.google.devtools.ksp.processing.Dependencies
+import com.google.devtools.ksp.processing.Resolver
+import com.google.devtools.ksp.processing.SymbolProcessor
+import com.google.devtools.ksp.processing.SymbolProcessorEnvironment
+import com.google.devtools.ksp.symbol.KSAnnotated
+import com.google.devtools.ksp.symbol.KSClassDeclaration
+import com.google.devtools.ksp.symbol.KSFile
+import com.google.devtools.ksp.validate
+import com.squareup.kotlinpoet.FileSpec
+
+object GlideSymbolProcessorConstants {
+  val PACKAGE_NAME: String = GlideSymbolProcessor::class.java.`package`.name
+  const val SINGLE_APP_MODULE_ERROR = "You can have at most one AppGlideModule, but found: %s"
+  const val DUPLICATE_LIBRARY_MODULE_ERROR =
+    "LibraryGlideModules %s are included more than once, keeping only one!"
+  const val INVALID_ANNOTATED_CLASS =
+    "@GlideModule annotated classes must implement AppGlideModule or LibraryGlideModule: %s"
+}
+
+private const val APP_MODULE_QUALIFIED_NAME = "com.bumptech.glide.module.AppGlideModule"
+private const val LIBRARY_MODULE_QUALIFIED_NAME =
+  "com.bumptech.glide.module.LibraryGlideModule"
+
+class GlideSymbolProcessor(private val environment: SymbolProcessorEnvironment) : SymbolProcessor {
+
+  override fun process(resolver: Resolver): List<KSAnnotated> {
+    val symbols = resolver.getSymbolsWithAnnotation("com.bumptech.glide.annotation.GlideModule")
+    val invalidSymbols = symbols.filter { !it.validate() }.toList()
+    val validSymbols = symbols.filter { it.validate() }.toList()
+    return try {
+      processChecked(resolver, symbols, validSymbols, invalidSymbols)
+    } catch (e: InvalidGlideSourceException) {
+      environment.logger.error(e.userMessage)
+      invalidSymbols
+    }
+  }
+
+  private fun processChecked(
+    resolver: Resolver,
+    symbols: Sequence<KSAnnotated>,
+    validSymbols: List<KSAnnotated>,
+    invalidSymbols: List<KSAnnotated>,
+  ): List<KSAnnotated> {
+    environment.logger.logging("Found symbols, valid: $validSymbols, invalid: $invalidSymbols")
+
+    val (appGlideModules, libraryGlideModules) = extractGlideModules(validSymbols)
+
+    if (libraryGlideModules.size + appGlideModules.size != validSymbols.count()) {
+      val invalidModules =
+        symbols.filter { !libraryGlideModules.contains(it) && !appGlideModules.contains(it) }
+          .map { it.location.toString() }
+          .toList()
+
+      throw InvalidGlideSourceException(
+        GlideSymbolProcessorConstants.INVALID_ANNOTATED_CLASS.format(invalidModules)
+      )
+    }
+
+    if (appGlideModules.size > 1) {
```

Comment:
KSP's multiple round logic will only feed processor with deferred symbols (return result from `SymbolProcessor.process()`) and newly generated files, therefore in case of a second round, if an annotated `AppGlideModule` is generated by other annotation processors (I doubt how likely it is though), then this check won't catch the fact that we have already processed `AppGlideModule` in previous round, as what you get from `Resolver` only take accounts for newly generated file and deferred symbols. 

Suggestion here is to use a state to store whether an `AppGlideModule` has been processed in `SymbolProcessor`, processors are only initialized once when loading, therefore it is fine to store state there.